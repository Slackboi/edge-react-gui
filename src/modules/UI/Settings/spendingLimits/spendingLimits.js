// @flow

import { combineReducers } from 'redux'
import { pathOr } from 'ramda'

// ACTIONS
import { ACCOUNT_INIT_COMPLETE } from '../../../../constants/indexConstants.js'

export const PREFIX = 'SPENDING_LIMITS/'

export const UPDATE = PREFIX + 'UPDATE'
export const update = (spendingLimits: SpendingLimits) => ({
  type: UPDATE,
  data: { spendingLimits }
})

// REDUCERS
export const initialState = {
  transaction: {
    isEnabled: true,
    amount: 100
  }
}
export type SpendingLimits = {
  transaction: {
    isEnabled: boolean,
    amount: number
  }
}

export const isEnabled = (state: boolean = true, action: Action) => {
  switch (action.type) {
    case ACCOUNT_INIT_COMPLETE: {
      const result = pathOr(state, ['spendingLimits', 'transaction', 'isEnabled'], action)
      return result
    }
    case UPDATE: {
      const result = pathOr(state, ['data', 'spendingLimits', 'transaction', 'isEnabled'], action)
      return result
    }
    default:
      return state
  }
}

export const amount = (state: number = 100, action: Action) => {
  switch (action.type) {
    case ACCOUNT_INIT_COMPLETE: {
      const result = pathOr(state, ['spendingLimits', 'transaction', 'amount'], action)
      return result
    }
    case UPDATE: {
      const result = pathOr(state, ['data', 'spendingLimits', 'transaction', 'amount'], action)
      return result
    }
    default:
      return state
  }
}

export const transaction = (state, action) => ({
  isEnabled: isEnabled(state.isEnabled, action),
  amount: amount(state.amount, action)
})

export const spendingLimits = (state, action) => ({
  transaction: transaction(state.transaction, action)
})
