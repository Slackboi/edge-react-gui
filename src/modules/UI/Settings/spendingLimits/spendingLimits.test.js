/* globals jest describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import { spendingLimits, initialState } from './spendingLimits.js'
import { ACCOUNT_INIT_COMPLETE } from '../../../../constants/indexConstants.js'
import { update } from './spendingLimits.js'

describe('spendingLimits', () => {
  it('should render initialState', () => {
    const actual = spendingLimits(undefined, {})

    expect(actual).toMatchSnapshot()
  })

  it('should update when logging in', () => {
    const loginAction = {
      type: ACCOUNT_INIT_COMPLETE,
      spendingLimits: {
        transaction: {
          isEnabled: false,
          amount: 150
        }
      }
    }
    const actual = spendingLimits(initialState, loginAction)

    expect(actual).toMatchSnapshot()
  })

  it('should use defaults when logging in, if not included in action', () => {
    const loginAction = {
      type: ACCOUNT_INIT_COMPLETE
    }
    const actual = spendingLimits(initialState, loginAction)

    expect(actual).toMatchSnapshot()
  })

  it('should update when receiving update spending limits action', () => {
    const updateAction = update({
      transaction: {
        isEnabled: false,
        amount: 234
      }
    })
    const actual = spendingLimits(initialState, updateAction)

    expect(actual).toMatchSnapshot()
  })
})
