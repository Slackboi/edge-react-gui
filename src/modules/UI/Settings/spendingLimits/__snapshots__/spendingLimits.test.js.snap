// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spendingLimits should render initialState 1`] = `
Object {
  "transaction": Object {
    "amount": 100,
    "isEnabled": true,
  },
}
`;

exports[`spendingLimits should update when logging in 1`] = `
Object {
  "transaction": Object {
    "amount": 150,
    "isEnabled": false,
  },
}
`;

exports[`spendingLimits should update when receiving update spending limits action 1`] = `
Object {
  "transaction": Object {
    "amount": 234,
    "isEnabled": false,
  },
}
`;

exports[`spendingLimits should use defaults when logging in, if not included in action 1`] = `
Object {
  "transaction": Object {
    "amount": 100,
    "isEnabled": true,
  },
}
`;
